### ISSUE TITLE
Change Request: Allow rules to specify how to merge options

### ISSUE THREAD

**Issue by matwilko** [eslint/eslint#17389]
### ESLint version

8.44.0

### What problem do you want to solve?

At the moment, flat config will nicely merge sets of rules together, and handle changing severities, but if rule options differ, later config elements simply override earlier ones (equivalent to `merge: "replace"`).

I think it would be worthwhile to allow rules to participate in the merging process - if two config elements need to merge differing options for a rule, then the rule should be able to decide how those options are merged.

The specific case that has motivated this is `eslint-plugin-putout` and their `putout/putout` rule, that has as its sole option a config object for `putout`, with an identical format to ESLint rule configs:

```
export default {
    plugins: { putout: require('eslint-plugin-putout') },
    rules: {
        "putout/putout": ["error", {
            rules: {
                'apply-template-literals': 'off',
                'remove-empty': 'off',
                'nodejs/remove-process-exit': 'off',
                'typescript/remove-unused-types': 'off',
                'remove-unused-variables': 'off',
                'remove-unused-expressions': 'off',
                'remove-unreachable-code': 'off',
                'remove-useless-arguments': 'off',
                'remove-useless-variables/rename': 'off',
                'remove-useless-return': 'off',
                'remove-useless-spread': 'off',
                'remove-unreferenced-variables': 'off',
                'tape/remove-skip': 'off',
                'tape/remove-only': 'off',
                'remove-console': 'off',
                'remove-debugger': 'off',
                'for-of/for': 'off',
                'for-of/remove-unused-variables': 'off',
                'for-of/remove-useless': 'off',
                'maybe/noop': 'off',
            },
        }]
    }
}
```

Given the way this particular rule is constructed, there's no way to selectively change the `putout` config without copy+pasting the entire option object, as the option object is entirely overridden on merge.

In my own codebase, I've had to write a post-processsor to do this search for any rule configurations of `putout/putout` and attempt to merge them properly ahead of time, but this isn't really possible when there are `file` scoped config elements present.

This could also be useful for the various `no-restricted-*` rules, where users can't currently merge multiple definitions together without manually copying.

### What do you think is the correct solution?

I think a reasonable solution would be to add a `rule.meta.mergeOptions` function to the rules spec.

The function would have the signature `function(options: unknown[], optionsToMerge: unknown[]): unknown[]`, allowing the rule to do any arbitrary merging logic it needs. The output would be sent through the rule options schema as a sanity check.

The function would only be called when there is a conflict during the usual config merge process, e.g. it would only be called when both first and second rule configurations specify some options, and those options are not already element-wise strictly equal (`===`).

e.g.
```
[
    {
        plugins: {
             customPlugin: {
                 rules: {
                     customRule: {
                         meta: {
                             mergeOptions(options, optionsToMerge) { console.log(`merging '${options}' and '${optionsToMerge}'); return [options + optionsToMerge]; }
                             schema: [{ type: "number" }]
                         },
                         create() {}
                     }
                 }
            }
        }
    },
    { rules: { "customPlugin/customRule": ["error"] } },     // ["error"]
    { rules: { "customPlugin/customRule": ["error", 1] } },  // ["error", 1]
    { rules: { "customPlugin/customRule": ["warn"] } },      // ["warn", 1]
    { rules: { "customPlugin/customRule": ["off", 2] } },    // ["off", 3] // "merging '1' and '2'"
    { rules: { "customPlugin/customRule": ["off"] } },       // ["off", 3]
    { rules: { "customPlugin/customRule": ["off", 3] } },    // ["off", 3]
    { rules: { "customPlugin/customRule": ["error", 4] } },  // ["error", 7] // "merging '3' and '4'"
```

Expected Output:
```
merging '1' and '2'
merging '3' and '4'
```

[rulesSchema.merge](https://github.com/eslint/eslint/blob/42faa17b1c93f801b14bea2840d1d528e25c7211/lib/config/flat-config-schema.js#L339) would need updating to use this new custom merging behaviour when available. If `rule.meta.mergeOptions` does not exist, then the current behaviour of simply replacing the options should prevail.

There's some complication here in that rules are usually only resolved once all merges have completed, and the `merge` function of each schema element is supposed to be independent (i.e. `rules` isn't supposed to depend on `plugins` during config resolution.)

I'm open to a different design that avoids having to delve down into `object-schema` and add functionality.

If adding functionality to `object-schema` is the preferred mechanism, I'd suggest the following:
- Add a `depends` key to the schema to allow keys to express a dependency on other keys (`require` isn't quite the right solution, we need to express "these keys need to have any processing (e.g. `merge`) done on them before this key is processed)
- If a schema has a `depends` key then `validate` and `merge` are modified such that:
  - Any key that is in the `depends` of another key is validated before that key.
  - When merging, any key that is in the `depends` of another key is merged before that key.
  - When merging, if the key has a `depends` key, an extra argument is passed to `merge` containing the currently merged config object, but with only the keys listed in `depends`. The method would be expected not to modify this object.

This solution would mean that it would no longer be possible to add a plugin with a named rule _after_ a config element referencing that rule, which is currently valid to do, but to be honest, that seems unlikely to happen in practice, and feels quite weird as a concept. 😄 It is a breaking change nonetheless.

(The breaking change could be softened by pre-scanning all plugin rules and determining if there are any mergeOptions functions, and if not, retain the old behaviour - this would _sort of_ push the breaking nature out into plugins who want to implement the feature - adding `mergeOptions` would be considered a breaking change for the plugin because it could break user configs if they upgrade the plugin. This feels a bit hostile to the ecosystem though!)

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
This has been discussed before:
https://github.com/eslint/eslint/issues/9192

There was extended discussion about the pros and cons, so I'd suggest reading through to get a ground level understanding of what the concerns were.

I'm not against revisiting this, but I do want to call out that this is addressing an edge case. Most rules will just want to use the default merging of options, so whatever solution we come up with should keep that in mind. Once again, we don't want to add a bunch of complexity for something that will have very few consumers.

**Comment by nzakas** [comment]
@mdjermanovic looking for your feedback on this.

**Comment by nzakas** [comment]
@matwilko sounds good. No rush!

**Comment by mdjermanovic** [comment]
I personally think that allowing each rule to have a different merging of options would be confusing for end users.

I'd lean towards the alternative suggested in https://github.com/eslint/eslint/issues/9192, which is to allow end users to specify how to merge options. With the new config system, it could be programmatically by providing a function that gets config for the rule calculated to that point and returns a new config, for example:

```js
{
    rules: {
        "no-restricted-syntax": config => config.concat("mySelector1", "mySelector2")
    }
}
```

**Comment by nzakas** [comment]
So in that case, we'd end up allowing a string, an array, or a function for each rule's configuration? 
And the function would be passed the rule config and be expected to return either a string or an array?

My only concern with that approach is that if the function returns an invalid configuration, it could be hard to track down where it was coming from, especially if you're using a shareable config.

**Comment by nzakas** [comment]
@mdjermanovic please see my previous comment for followup questions.

**Comment by mdjermanovic** [comment]
> So in that case, we'd end up allowing a string, an array, or a function for each rule's configuration?
> And the function would be passed the rule config and be expected to return either a string or an array?

Yes.

Alternatively (or additionally), `rules` could be a function that gets the whole rules object calculated to that point and returns a new object. That could also solve other problems, like disabling all rules from a plugin (https://github.com/eslint/eslint/issues/14026) and changing all severities (https://github.com/eslint/eslint/issues/14679).

**Comment by nzakas** [comment]
Interesting points. I think we're now at the point where we'd need an RFC to go any further down this track, as this is sufficiently complex that tracking details throughout comments isn't going to get us very far.

**Comment by Rec0iL99** [comment]
> Interesting points. I think we're now at the point where we'd need an RFC to go any further down this track, as this is sufficiently complex that tracking details throughout comments isn't going to get us very far.

@nzakas should we push this forward for an RFC?

**Comment by nzakas** [comment]
@Rec0iL99 yes, sir!

**Comment by Rec0iL99** [comment]
Not stale. Waiting for an RFC.

cc @matwilko

**Comment by matwilko** [comment]
Hey folks, apologies for the radio silence - I'd be happy to put together an RFC for this, but it's unfortunately a level of contribution that would need some approvals from my employer to get started on (I can only make "trivial" contributions on my own recognisance!)

If you're all willing to continue being patient with me, I'll work on getting things pushed through here, but I wouldn't want it to wait on me if somebody else wants to take it up in the meantime :)

**Comment by nzakas** [comment]
We don't have any set schedule, so take your time. 

I'd just ask you post here again if it ends up taking longer than a month to let us know if you're still working towards getting approval.

**Comment by Rec0iL99** [comment]
Not stale. Waiting for an RFC.

@matwilko are still working on this?

**Comment by abdatta** [comment]
Eagerly waiting for an update on this! I have a similar need for the `no-restricted-properties` option, where base config options are getting overridden rather than extended.

**Comment by nzakas** [comment]
@abdatta please don't leave comments like this. They create unnecessary noise for the maintainers. Updates will be posted to this issue, and if you have an interest, I'd suggest you might want to open an RFC yourself.
